R/pmcmc_tools.R

```diff
@@ -194,3 +194,108 @@ combine_state <- function(x) {
   ret$state <- state
   ret
 }
+
+
+##' Convert sampled trajectories from [pmcmc()] to a Tidy (long) format,
+##' and optionally marginalise over particles to produce summary statistics.
+##' Helper function useful for ggplotters.
+##'
+##' @title Convert sampled trajectories from [pmcmc()] to a Tidy (long) format
+##' @param object Results of running [pmcmc()].
+##'
+##' @param summarise Logical indicating whether the trajectories should be
+##' marginalised over particles to produce summary statistics. Default is
+##' `FALSE`.
+##'
+##' @param quantiles Optional vector giving quantiles to output if `summarise`
+##' is TRUE defaults to c(0.025, 0.5, 0.975) corresponding to 2.5-, 50-, and
+##' 97.5-percentiles
+##'
+##' @param summary_function Optional function to calculated user-defined
+##' summary statistics - defaults to calculating mean, 2.5-, 50-, and
+##' 97.5-percentiles
+##'
+##' @export
+##' @importFrom stats quantile
+
+pmcmc_tidy_trajectories <- function(object, summarise = FALSE,
+                                    quantiles = c(0.025, 0.5, 0.975),
+                                    summary_function = NULL) {
+
+  assert_is(object, "mcstate_pmcmc")
+  assert_scalar_logical(summarise)
+
+  state <- object$trajectories$state
+
+  # create a named list with an entry for each dimension of `state`
+  # the list entries will be dimnames, where they exist, and numeric otherwise
+  grid <- list(state = rownames(state) %||% seq_len(nrow(state)))
+  if (object$nested) grid$population <- colnames(state)
+  if (summarise) {
+    # define summary_function and a vector of output names
+    f <- summary_function %||% pmcmc_summarise_one(quantiles)
+    summary_output <- f(numeric(1))
+    grid$statistic <- names(summary_output) %||% seq_along(summary_output)
+    } else {
+    grid$particle <- seq_len(nrow(object$pars))
```

Comment:
I am a bit confused about what is going on here tbh. I think we should avoid the `f(numeric(1))` though and look at what was created in the subsequent block